sapi/fpm/fpm/fpm_sockets.c

```diff
@@ -60,38 +61,36 @@ static void fpm_sockets_cleanup(int which, void *arg) /* {{{ */
 	unsigned socket_set[FPM_ENV_SOCKET_SET_MAX];
 	unsigned socket_set_buf = 0;
 	char envname[32];
-	char *env_value = 0;
-	int p = 0;
+	smart_string env_str = {0};
 	struct listening_socket_s *ls = sockets_list.data;
 
 	for (i = 0; i < sockets_list.used; i++, ls++) {
 		if (which != FPM_CLEANUP_PARENT_EXEC) {
 			close(ls->sock);
 		} else { /* on PARENT EXEC we want socket fds to be inherited through environment variable */
 			char fd[32];
-			char *tmpenv_value;
 			snprintf(fd, sizeof(fd), "%d", ls->sock);
 
 			socket_set_buf = (i % FPM_ENV_SOCKET_SET_SIZE == 0 && i) ? 1 : 0;
-			tmpenv_value = realloc(env_value, p + (p ? 1 : 0) + strlen(ls->key) + 1 + strlen(fd) + socket_set_buf + 1);
-			if (!tmpenv_value) {
-				zlog(ZLOG_SYSERROR, "failure to inherit data on parent exec for socket `%s` due to memory allocation failure", ls->key);
-				free(ls->key);
-				break;
-			}
-
-			env_value = tmpenv_value;
 
 			if (i % FPM_ENV_SOCKET_SET_SIZE == 0) {
-				socket_set[socket_set_count] = p + socket_set_buf;
+				socket_set[socket_set_count] = env_str.len + socket_set_buf;
 				socket_set_count++;
 				if (i) {
-					*(env_value + p + 1) = 0;
+					smart_string_appendc(&env_str, '\0');
 				}
 			}
 
-			p += sprintf(env_value + p + socket_set_buf, "%s%s=%s", (p && !socket_set_buf) ? "," : "", ls->key, fd);
-			p += socket_set_buf;
+			if (env_str.len && !socket_set_buf) {
+				smart_string_appendc(&env_str, ',');
+			}
+			smart_string_appends(&env_str, ls->key);
+			smart_string_appendc(&env_str, '=');
+			smart_string_appends(&env_str, fd);
```

Comment:
Can this be `smart_string_append_printf` or `smart_string_append_long` to avoid the manual `snprintf()` at the top?